cmd/lifecycle/builder.go

```diff
@@ -48,26 +60,160 @@ func (b *buildCmd) Args(nargs int, args []string) error {
 }
 
 func (b *buildCmd) Privileges() error {
-	// builder should never be run with privileges
-	if priv.IsPrivileged() {
+	_, privGroup, _, err := b.readData()
+	if err != nil {
+		return err
+	}
+
+	hasPrivGroups := len(privGroup.Group) > 0
+
+	// builder should never be run with privileges if there aren't any stack buildpacks
+	if !hasPrivGroups && priv.IsPrivileged() {
 		return cmd.FailErr(errors.New("refusing to run as root"), "build")
+	} else if hasPrivGroups && !priv.IsPrivileged() {
+		return cmd.FailErr(errors.New("must run as root"), "build")
 	}
 	return nil
 }
 
 func (b *buildCmd) Exec() error {
-	group, plan, err := b.readData()
+	group, privGroup, plan, err := b.readData()
+	if err != nil {
+		return err
+	}
+
+	if err := verifyBuildpackApis(group, privGroup); err != nil {
+		return err
+	}
+
+	if len(privGroup.Group) == 0 {
+		return b.build(group, plan)
+	}
+	return b.buildWithReexec(group, privGroup, plan)
+}
+
+func (ba buildArgs) buildAll(group, privGroup lifecycle.BuildpackGroup, plan lifecycle.BuildPlan, rootuid, rootgid, uid, gid int) error {
+	if len(privGroup.Group) > 0 {
+		if err := priv.RunAsEffective(rootuid, rootgid); err != nil {
+			return err
+		}
+		if err := ba.stackBuild(privGroup, plan); err != nil {
+			return err
+		}
+		if err := priv.RunAsEffective(uid, gid); err != nil {
+			return err
+		}
+	}
+
+	return ba.build(group, plan)
+}
+
+func (ba buildArgs) stackBuild(privGroup lifecycle.BuildpackGroup, plan lifecycle.BuildPlan) error {
+	stackSnapshotter, err := snapshot.NewKanikoSnapshotter("/", ba.layersDir, ba.platformDir)
 	if err != nil {
 		return err
 	}
-	if err := verifyBuildpackApis(group); err != nil {
+
+	workspaceDir, err := ioutil.TempDir("", "stack-workspace")
+	if err != nil {
+		return err
+	}
+
+	builder, err := ba.createBuilder(privGroup, plan)
+	if err != nil {
 		return err
 	}
-	return b.build(group, plan)
+
+	builder.Snapshotter = stackSnapshotter
+	builder.AppDir = workspaceDir
+	builder.BuildpacksDir = ba.stackBuildpacksDir
+
+	_, err = ba.execBuild(builder)
+	return err
 }
 
 func (ba buildArgs) build(group lifecycle.BuildpackGroup, plan lifecycle.BuildPlan) error {
-	builder := &lifecycle.Builder{
+	builder, err := ba.createBuilder(group, plan)
```

Comment:
Instead of creating two separate builders for stack and regular builds I think we should be able to execute both parts of the build with the same builder struct.